server/sonar-server/src/main/java/org/sonar/server/computation/task/projectanalysis/issue/IssueCreationDateCalculator.java

```
@@ -0,0 +1,123 @@
+import org.sonar.server.computation.task.projectanalysis.analysis.AnalysisMetadataHolder;
+import org.sonar.server.computation.task.projectanalysis.component.Component;
+import org.sonar.server.computation.task.projectanalysis.qualityprofile.ActiveRule;
+import org.sonar.server.computation.task.projectanalysis.qualityprofile.ActiveRulesHolder;
+import org.sonar.server.computation.task.projectanalysis.scm.Changeset;
+import org.sonar.server.computation.task.projectanalysis.scm.ScmInfo;
+import org.sonar.server.computation.task.projectanalysis.scm.ScmInfoRepository;
+import org.sonar.server.issue.IssueFieldsSetter;
+
+import static org.sonar.core.issue.IssueChangeContext.createScan;
+
+/**
+ * Calculates the creation date of an issue. Takes into account, that the issue
+ * might be raised by adding a rule to a quality profile.
+ */
+public class IssueCreationDateCalculator extends IssueVisitor {
+
+  private static final Logger LOGGER = Loggers.get(IssueCreationDateCalculator.class);
+
+  private final ScmInfoRepository scmInfoRepository;
+  private final IssueFieldsSetter issueUpdater;
+  private final AnalysisMetadataHolder analysisMetadataHolder;
+  private final IssueChangeContext changeContext;
+  private final ActiveRulesHolder activeRulesHolder;
+
+  public IssueCreationDateCalculator(AnalysisMetadataHolder analysisMetadataHolder, ScmInfoRepository scmInfoRepository,
+    IssueFieldsSetter issueUpdater, ActiveRulesHolder activeRulesHolder) {
+    this.scmInfoRepository = scmInfoRepository;
+    this.issueUpdater = issueUpdater;
+    this.analysisMetadataHolder = analysisMetadataHolder;
+    this.changeContext = createScan(new Date(analysisMetadataHolder.getAnalysisDate()));
+    this.activeRulesHolder = activeRulesHolder;
+  }
+
+  @Override
+  public void onIssue(Component component, DefaultIssue issue) {
+    if (issue.isNew() && ruleIsNew(issue)) {
+
+      // get scm information of file, if available
```

hint of this being abusive: 
when there is no changeset, date of issue is not altered... this is hidden in the last line `ifPresent` while for understanding by reader, this should be written as a fail fast condition

functional programming is about making code closer to the algorithm... this is quite the opposite